host-tools/callisto-lib/src/lib.rs

```diff
@@ -0,0 +1,327 @@
+        cur_wait_time: u32,
+    },
+}
+
+const _: () = assert!(12 == size_of::<ReframingState>());
+const _: () = assert!(4 == align_of::<ReframingState>());
+
+const _: () = assert!(
+    0u32 == {
+        let v = ReframingState::Detect;
+        unsafe { *(&v as *const _ as *const u32) }
+    }
+);
+
+const _: () = assert!(
+    1u32 == {
+        let v = ReframingState::Done;
+        unsafe { *(&v as *const _ as *const u32) }
+    }
+);
+
+#[allow(dead_code)]
+const SOME_RF_STATE_WAIT: ReframingState = ReframingState::Wait {
+    target_correction: 3.321,
+    cur_wait_time: 12345,
+};
+
+#[allow(dead_code)]
+const CASTED_RF_STATE_WAIT: *const (u32, f32, u32) =
+    &SOME_RF_STATE_WAIT as *const _ as *const (u32, f32, u32);
+
+const _: () = assert!(
+    2u32 == {
+        let (x, _, _) = unsafe { *CASTED_RF_STATE_WAIT };
+        x
+    }
+);
+
+const _: () = assert!(
+    3.321f32 == {
+        let (_, x, _) = unsafe { *CASTED_RF_STATE_WAIT };
+        x
+    }
+);
+
+const _: () = assert!(
+    12345u32 == {
+        let (_, _, x) = unsafe { *CASTED_RF_STATE_WAIT };
+        x
+    }
+);
+
+/// Rust sibling of `Bittide.ClockControl.Callisto.Types.ControlSt`.
+#[repr(C)]
+pub struct ControlSt {
+    /// Accumulated speed change requests, where
+    /// * `speedup ~ 1`
+    /// * `slowdown ~ -1`
+    pub z_k: i32,
+    /// Previously submitted speed change request. Used to determine
+    /// the estimated clock frequency.
+    pub b_k: SpeedChange,
+    /// Steady-state value (determined when stability is detected for
+    /// the first time).
+    pub steady_state_target: f32,
+    /// finite state machine for reframing detection
+    pub rf_state: ReframingState,
+}
+
+const _: () = assert!(24 == size_of::<ControlSt>());
+const _: () = assert!(4 == align_of::<ControlSt>());
+
+#[allow(dead_code)]
+const SOME_CONTROL_STATE: ControlSt = ControlSt {
+    z_k: 54321,
+    b_k: SpeedChange::SlowDown,
+    steady_state_target: 3.185,
+    rf_state: SOME_RF_STATE_WAIT,
+};
+
+#[allow(dead_code)]
+const CASTED_CONTROL_STATE: *const (i32, u32, f32, u32, f32, u32) =
+    &SOME_CONTROL_STATE as *const _ as *const (i32, u32, f32, u32, f32, u32);
+
+const _: () = assert!(
+    54321i32 == {
+        let (x, _, _, _, _, _) = unsafe { *CASTED_CONTROL_STATE };
+        x
+    }
+);
+
+const _: () = assert!(
+    1u32 == {
+        let (_, x, _, _, _, _) = unsafe { *CASTED_CONTROL_STATE };
+        x
+    }
+);
+
+const _: () = assert!(
+    3.185f32 == {
+        let (_, _, x, _, _, _) = unsafe { *CASTED_CONTROL_STATE };
+        x
+    }
+);
+
+const _: () = assert!(
+    2u32 == {
+        let (_, _, _, x, _, _) = unsafe { *CASTED_CONTROL_STATE };
+        x
+    }
+);
+
+const _: () = assert!(
+    3.321f32 == {
+        let (_, _, _, _, x, _) = unsafe { *CASTED_CONTROL_STATE };
+        x
+    }
+);
+
+const _: () = assert!(
+    12345u32 == {
+        let (_, _, _, _, _, x) = unsafe { *CASTED_CONTROL_STATE };
+        x
+    }
```

Comment:
This could be done like this, right? Should do the same thing but not be split over so many different assertions? But that's just a small "beauty" suggestion :sweat_smile: 

```suggestion
const _: () = assert!(
    {
        let fields@(_, _, _, _, _, _) = unsafe { *CASTED_CONTROL_STATE };
        fields == (54321i32, 1u32, 3.185f32, 2u32, 3.321f32, 12345u32)
    }
);
```

Reply:
It could, but then you don't  know any more which field was wrong if the assertion fails. Unfortunately, the compiler does not give you that information. Therefore, I usually prefer one assertion per property.

Reply:
That's a fair point, we can keep it that way then! :)